File: movie-infrastructures/storage/src/main/java/com/movie/storage/movie/repository/ReservationRepositoryImpl.java
package com.movie.storage.movie.repository;

import com.movie.domain.movie.ReservationRepository;
import com.movie.domain.movie.domain.Reservation;
import com.movie.domain.movie.dto.command.ReservationCommand;
import com.movie.storage.mapper.ModelMapper;
import com.movie.storage.movie.dto.payload.ReservationPayload;
import com.movie.storage.movie.entity.ReservationEntity;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
@RequiredArgsConstructor
public class ReservationRepositoryImpl implements ReservationRepository {

    private final ReservationJpaQuerySupport reservationJpaQuerySupport;

    @Override
    public List<Reservation> getReservations(ReservationCommand.Get get) {
        List<ReservationPayload.Get> reservations = reservationJpaQuerySupport.getReservations(get.schedule(), get.seatIds());
        return reservations.stream().map(ModelMapper.ReservationMapper::fromPayload).toList();
    }

    @Override
    public void makeReservation(ReservationCommand.Reserve reserve) {
        List<ReservationEntity> reservations = reservationJpaQuerySupport.getReservationEntities(reserve.scheduleId(), reserve.seatIds());
        reservations.forEach(reservation -> {
            reservation.makeReservation(reserve.userId());
        });
    }

}

File: movie/infra/src/main/kotlin/com/example/movie/persistence/reservation/repository/ReservationRepositoryImpl.kt
package com.example.movie.persistence.reservation.repository

import com.example.movie.domain.reservation.model.Reservation
import com.example.movie.domain.reservation.repository.ReservationRepository
import com.example.movie.domain.screening.model.Screening
import com.example.movie.domain.seat.model.Seat
import com.example.movie.persistence.reservation.entity.ReservationEntity
import org.springframework.stereotype.Repository

@Repository
class ReservationRepositoryImpl(
    private val reservationJpaRepository: ReservationJpaRepository
) : ReservationRepository {
    override fun findByScreeningAndSeats(screening: Screening, seats: List<Seat>) : List<Reservation> {
        val reservations = reservationJpaRepository.findByScreeningIdAndSeatIds(
            screeningId = screening.id,
            seatIds = seats.map { it.id })
        return reservations.map { it.toDomain() }
    }

    override fun findByScreeningIdAndUserId(screeningId: Long, userId: Long): List<Reservation> {
        return reservationJpaRepository.findByScreeningIdAndUserId(screeningId, userId).map { it.toDomain() }
    }

    override fun saveAll(reservations: List<Reservation>) : List<Reservation> {
        return reservationJpaRepository.saveAll(reservations.map{ ReservationEntity.from(it)}).map { it.toDomain() }
    }
}